lib/Parse/ParseExpr.cpp

```diff
@@ -3756,6 +3756,162 @@ ParserResult<Expr> Parser::parseExprGradientBody(ExprKind kind) {
   return makeParserResult<Expr>(result);
 }
 
+/// SWIFT_ENABLE_TENSORFLOW
+/// parseFunctionIdentifier
+///
+///   function-identifier:
+///     type-identifier? unqualified-decl-name
+///   type-identifier:
+///     identifier generic-args? ('.' identifier generic-args?)*
+///
+/// Note: this function is largely copied from `parseTypeIdentifier` and should
+/// be kept in sync.
```

Comment:
I'll make `parseTypeIdentifier` more general.
Changes to `parseTypeIdentifier` will be a bit invasive, but I think it's worth it because the level of dupe is not acceptable and `parseQualifiedDeclName` is a generally useful function.

I'll name the flag `isParsingQualifiedDeclName` for now: open to name suggestions.